### PR TITLE
Capturable Adam optimizer

### PR DESCRIPTION
**Pull request by gilshm**
This PR adds a "capturable" argument to enable the Adam optimizer to be captured by CUDA graphs. This enhancement also supports non-static LR strategies.

### CODE REVIEW

File: csrc/multi_tensor_adam.cu [hunk 1]
```diff
@@ -126,6 +126,118 @@ struct AdamFunctor
...
+    TensorListMetadata<4>& tl,
+    const float beta1,
+    const float beta2,
+    const int* step,
+    const int bias_correction,
+    const float epsilon,
+    const float* lr,
+    adamMode_t mode,
+    const float decay,
+    const float* inv_scale)
```
Comment (crcrpar) [inline csrc/multi_tensor_adam.cu:144]: Is this the argument which differentiates this functor from the existing `AdamFunctor`?

File: apex/optimizers/fused_adam.py [hunk 2]
```diff
@@ -150,44 +166,104 @@ def step(self, closure=None, grads=None, output_params=None, scale=None, grad_no
...
+                        group['lr'],
+                        beta1,
+                        beta2,
+                        group['eps'],
+                        group['step'],
+                        self.adam_w_mode,
+                        bias_correction,
+                        group['weight_decay'],
+                    )
+                if(len(g_32) > 0):
```
Comment (crcrpar) [inline apex/optimizers/fused_adam.py:256]: nit-picking
```suggestion
                if len(g_32) > 0:
```

or 
```suggestion
                if g_32:
```

File: apex/optimizers/fused_adam.py [hunk 3]
```diff
@@ -150,44 +166,104 @@ def step(self, closure=None, grads=None, output_params=None, scale=None, grad_no
...
+					 beta1,
+					 beta2,
+					 group['eps'],
+					 group['step'],
+					 self.adam_w_mode,
+					 bias_correction,
+					 group['weight_decay'],
+					 inv_scale)
+            else:
+                if(len(g_16) > 0):
```
Comment (crcrpar) [inline apex/optimizers/fused_adam.py:230]: nit-picking
```suggestion
                if len(g_16) > 0:
```
or
```suggestion
                if g_16:
```

File: apex/optimizers/fused_adam.py [hunk 4]
```diff
@@ -150,44 +166,104 @@ def step(self, closure=None, grads=None, output_params=None, scale=None, grad_no
...
+					 group['lr'],
+					 beta1,
+					 beta2,
+					 group['eps'],
+					 group['step'],
+					 self.adam_w_mode,
+					 bias_correction,
+					 group['weight_decay'],
+					 inv_scale)
+            else:
```
Comment (crcrpar) [inline apex/optimizers/fused_adam.py:229]: the indentation doesn't look consistent. could you apply some formatter?

File: apex/optimizers/fused_adam.py [hunk 5]
```diff
@@ -62,20 +62,35 @@ class FusedAdam(torch.optim.Optimizer):
...
-                 weight_decay=0., amsgrad=False, set_grad_none=True):
+                 weight_decay=0., amsgrad=False, set_grad_none=True,
+                 capturable=False):
 
         if amsgrad:
             raise RuntimeError('FusedAdam does not support the AMSGrad variant.')
+        # If the optimizer is capturable then LR should be a tensor (on GPU)
+        lr = torch.tensor(lr, dtype=torch.float32) if capturable else lr
         defaults = dict(lr=lr, bias_correction=bias_correction,
                         betas=betas, eps=eps, weight_decay=weight_decay)
```
Comment (crcrpar) [inline apex/optimizers/fused_adam.py:73]: I'm wondering if `load_state_dict` works as is when `capturable=True`

File: apex/contrib/test/optimizers/test_adam.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
+import copy
```
Comment (crcrpar) [inline apex/contrib/test/optimizers/test_adam.py:1]: could you put this file into tests/L0/run_optimizers?

File: apex/contrib/test/optimizers/test_adam.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+        self.model_ = Model().cuda()
+        self.model_.load_state_dict(copy.deepcopy(self.model.state_dict()))
+
+        self.lr = 0.00001
+        params = [p for p in self.model.parameters() if p.requires_grad]
+        self.optimizer = torch.optim.Adam(params, lr=self.lr)
+
+    def testGradScaler(self):
+        params_ = [p for p in self.model_.parameters() if p.requires_grad]
+        self.optimizer_ = apex.optimizers.FusedAdam(params_, lr=self.lr, capturable=False)
```
Comment (crcrpar) [inline apex/contrib/test/optimizers/test_adam.py:67]: [trivial question] why making the optimizer an attribute of this class?

Comment (crcrpar) [inline apex/contrib/test/optimizers/test_adam.py:67]: what do you think about testing both AdamW and Adam?

File: apex/contrib/test/optimizers/test_adam.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+
+            # Init for next iteration
+            self.optimizer.zero_grad()
+            self.optimizer_.zero_grad()
+
+            self.model_.load_state_dict(copy.deepcopy(self.model.state_dict()))
+    
+    def testGradScalerCapturable(self):
+        params_ = [p for p in self.model_.parameters() if p.requires_grad]
+        self.optimizer_ = apex.optimizers.FusedAdam(params_, lr=self.lr, capturable=True)
```
Comment (crcrpar) [inline apex/contrib/test/optimizers/test_adam.py:109]: ditto

File: apex/contrib/test/optimizers/test_adam.py [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+
+            # Init for next iteration
+            self.optimizer.zero_grad()
+            self.optimizer_.zero_grad()
+
+            self.model_.load_state_dict(copy.deepcopy(self.model.state_dict()))
+    
+    def testNative(self):
+        params_ = [p for p in self.model_.parameters() if p.requires_grad]
+        self.optimizer_ = apex.optimizers.FusedAdam(params_, lr=self.lr, capturable=False)
```
Comment (crcrpar) [inline apex/contrib/test/optimizers/test_adam.py:151]: ditto

File: apex/contrib/test/optimizers/test_adam.py [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+
+            scaler_.scale(loss_).backward()
+            scaler_.step(self.optimizer_)
+            scaler_.update()
+
+            for module in zip(self.model.modules(), self.model_.modules()):
+                m = module[0]
+                m_ = module[1]
+                if isinstance(m, nn.Conv2d) or isinstance(m_, nn.Linear):
+                    self.assertTrue(torch.allclose(m.weight, m_.weight, atol=1e-3, rtol=1e-3, equal_nan=True))
```
Comment (crcrpar) [inline apex/contrib/test/optimizers/test_adam.py:99]: nit-picking
```suggestion
                    torch.testing.assert_close(m.weight, m_.weight, atol=1e-3, rtol=1e-3, equal_nan=True)
```

File: apex/contrib/test/optimizers/test_adam.py [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
```
Comment (crcrpar) [inline apex/contrib/test/optimizers/test_adam.py:141]: ditto
nit-picking
```suggestion
                    torch.testing.assert_close(m.weight, m_.weight, atol=1e-3, rtol=1e-3, equal_nan=True)
```

File: apex/contrib/test/optimizers/test_adam.py [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+            loss_ = ((gt_ - y) ** 2).mean()
+
+            loss_.backward()
+            self.optimizer_.step()
+
+            for module in zip(self.model.modules(), self.model_.modules()):
+                m = module[0]
+                m_ = module[1]
+                if isinstance(m, nn.Conv2d) or isinstance(m_, nn.Linear):
+                    self.assertTrue(torch.allclose(m.weight, m_.weight, atol=1e-3, rtol=1e-3, equal_nan=True))
```
Comment (crcrpar) [inline apex/contrib/test/optimizers/test_adam.py:177]: ditto, nit-picking
```suggestion
                    torch.testing.assert_close(m.weight, m_.weight, atol=1e-3, rtol=1e-3, equal_nan=True)
```

File: apex/optimizers/fused_adam.py (hunk 5 above)
Comment (crcrpar) [reply apex/optimizers/fused_adam.py:73]: ah, there exists the check

File: csrc/multi_tensor_adam.cu [hunk 13]
```diff
@@ -169,3 +281,42 @@ void multi_tensor_adam_cuda(
...
+      AdamCapturableFunctor<scalar_t_0>(),
+      beta1,
+      beta2,
+      step.DATA_PTR<int>(),
+      bias_correction,
+      epsilon,
+      lr.DATA_PTR<float>(),
+      (adamMode_t) mode,
+      weight_decay,
+      inv_scale.data_ptr<float>()); )
```
Comment (crcrpar) [inline csrc/multi_tensor_adam.cu:317]: IMHO, it's okay to not use `DATA_PTR`

File: csrc/multi_tensor_adam.cu (hunk 1 above)
Comment (gilshm) [reply csrc/multi_tensor_adam.cu:144]: This functor is different from the original with the following:
1) const int* step - pointer, since it is now being used inside the kernel
2) const int bias_correction - same, see [here](https://github.com/NVIDIA/apex/blob/6588f6b03cf60ba65941d236a19dd5d046e7f90a/csrc/multi_tensor_adam.cu#L256-L261)
3) const float* lr - pointer, to support LR scheduling while captured
4) const float* inv_scale - same

File: apex/contrib/test/optimizers/test_adam.py (hunk 6 above)
Comment (gilshm) [reply apex/contrib/test/optimizers/test_adam.py]: Done

File: apex/contrib/test/optimizers/test_adam.py (hunk 7 above)
Comment (gilshm) [reply apex/contrib/test/optimizers/test_adam.py]: Fixed

Comment (gilshm) [reply apex/contrib/test/optimizers/test_adam.py]: You are right, it is not really necessary. I modified that so the tested optimizer is _not_ an attribute of the class.

Comment (gilshm) [reply apex/contrib/test/optimizers/test_adam.py]: I don't mind adding that to the testbench.
Does it only requires me to set `weight_decay` to non-zero value?

File: apex/contrib/test/optimizers/test_adam.py (hunk 10 above)
Comment (gilshm) [reply apex/contrib/test/optimizers/test_adam.py]: Fixed

File: apex/contrib/test/optimizers/test_adam.py (hunk 8 above)
Comment (gilshm) [reply apex/contrib/test/optimizers/test_adam.py]: Fixed

File: apex/contrib/test/optimizers/test_adam.py (hunk 11 above)
Comment (gilshm) [reply apex/contrib/test/optimizers/test_adam.py]: Fixed

File: apex/contrib/test/optimizers/test_adam.py (hunk 9 above)
Comment (gilshm) [reply apex/contrib/test/optimizers/test_adam.py]: Fixed

File: apex/contrib/test/optimizers/test_adam.py (hunk 12 above)
Comment (gilshm) [reply apex/contrib/test/optimizers/test_adam.py]: Fixed

File: apex/optimizers/fused_adam.py (hunk 2 above)
Comment (gilshm) [reply apex/optimizers/fused_adam.py]: Fixed

File: apex/optimizers/fused_adam.py (hunk 3 above)
Comment (gilshm) [reply apex/optimizers/fused_adam.py]: Fixed

File: apex/optimizers/fused_adam.py (hunk 4 above)
Comment (gilshm) [reply apex/optimizers/fused_adam.py]: Fixed

File: csrc/multi_tensor_adam.cu (hunk 13 above)
Comment (gilshm) [reply csrc/multi_tensor_adam.cu:317]: The thing is `lr` and `step` are `at::Tensor` (see [here](https://github.com/NVIDIA/apex/blob/930bfb59b343b7dc2dccb0ee9fd29e41a54b9853/csrc/multi_tensor_adam.cu#L285-L297)). Will it still be correct if I will remove `DATA_PTR`?

Comment (crcrpar) [reply csrc/multi_tensor_adam.cu:317]: oh I meant to say it's safe to use `data_ptr` instead like `inv_scale` as per https://github.com/NVIDIA/apex/blob/75f401e088ef88e7c85a57ecf70fb232235f0334/csrc/compat.h#L5-L9 (where `VERSION_GE_1_3` means the available pytorch's version >= 1.3)

Comment (gilshm) [reply csrc/multi_tensor_adam.cu]: Got it. I see I also used the lowercase version in line 317...
I modified everything to be lowercase as you suggested.

File: csrc/multi_tensor_adam.cu [hunk 14]
```diff
@@ -126,6 +126,118 @@ struct AdamFunctor
...
+      }
+#pragma unroll
+      for(int ii = 0; ii < ILP; ii++)
+      {
+        int i = i_start + threadIdx.x + ii*blockDim.x;
+        if(i < n && i < chunk_size)
+        {
+          p[i] = r_p[ii];
+          m[i] = r_m[ii];
+          v[i] = r_v[ii];
```
Comment (crcrpar) [inline csrc/multi_tensor_adam.cu:234]: just to confirm, as this block doesn't seem to store the gradients, wouldn't you need unscaled gradient values outside of `optimizer.step` when GradientScaler is used?

**Review by crcrpar — COMMENTED** [review]
I got a quick question

File: csrc/multi_tensor_adam.cu (hunk 14 above)
Comment (gilshm) [reply csrc/multi_tensor_adam.cu:234]: Perhaps I'm missing something... why do I need to store the gradients? as far as I understand, when gradient scaling is used and if overflow is detected, then the current iteration gradients are discarded. Doesn't it work like that?

Comment (crcrpar) [reply csrc/multi_tensor_adam.cu:234]: while non-fused Adam such as `torch.optim.Adam` would unscale gradients (technically `GradScaler` unscales the registered gradients on behalf) in the in-place manner, this optimizer doesn't. So if we're seriously interested in the parity between the native Adam and the fused one here, it would make sense to store the unscaled gradients

Comment (gilshm) [reply csrc/multi_tensor_adam.cu:235]: Fixed and added check to test.
